pyrex.py

```diff
@@ -665,35 +665,60 @@ def create_shims(config, build_config, buildconf):
     os.chmod(rebuildfile, stat.S_IRWXU)
 
     # Create shims
-    user_commands = config["config"].get("commands")
-    if user_commands:
-        user_commands = user_commands.split()
-        command_globs = [c for c in user_commands if not c.startswith("!")]
-        nopyrex_globs = [c[1:] for c in user_commands if c.startswith("!")]
-    else:
-        command_globs = build_config["container"].get("commands", {}).get("include", {})
-        nopyrex_globs = build_config["container"].get("commands", {}).get("exclude", {})
-
     commands = set()
 
-    def add_commands(globs, target):
+    def glob_commands(globs):
         nonlocal commands
 
         for g in globs:
             for cmd in glob.iglob(g):
-                norm_cmd = os.path.normpath(cmd)
+                base = os.path.basename(cmd)
                 if (
-                    norm_cmd not in commands
+                    base not in commands
                     and os.path.isfile(cmd)
                     and os.access(cmd, os.X_OK)
                 ):
-                    commands.add(norm_cmd)
-                    name = os.path.basename(cmd)
+                    commands.add(base)
+                    shim = os.path.join(shimdir, base)
+                    yield cmd, shim
+
+    # User commands
+    user_commands = config["config"].get("commands", "").split()
+
+    user_excludes = [c[1:] for c in user_commands if c.startswith("!")]
+    for cmd, shim in glob_commands(user_excludes):
+        os.symlink(cmd, shim)
+
+    user_includes = [c for c in user_commands if not c.startswith("!")]
+    for cmd, shim in glob_commands(user_includes):
+        # User commands need dedicates shims that execute with the full path,
+        # since we don't know if they are in $PATH
+        with open(shim, "w") as f:
+            f.write(
+                textwrap.dedent(
+                    """\
```

Comment:
Nit: this seems weird. Why is the opening line here indented even further than the others? I don't think the function requires that.

Reply:
It doesn't. `black` indents the `"""` line but leaves the rest so sometimes the indentation gets off when I refactor things.